Support publishing of Javadoc and sources

jjohannes opened the pull request:
Implementation of #10737 including:

## Production code (Java plugins)

- Each feature variant that is based on a separate source set now provides a `javadoc` task similar to the main feature (in analogy to other tasks like `compileJava`).
- The java extension now offers `publishJavadoc()` and `publishSources()` option - providing the corresponding `jar` task and registering the docs/sources as a variant used in publishing.
- The same option are available on `FeatureSpec` to configure the same for each feature variant.
- A compatibility rule for the `Category` attribute has been added to allow some useful resolution (although the focus of this PR is publishing)

## Documentation / samples / tests

- All samples/tests that individually created sourcesJar or javadocJar tasks have been updated to use the new feature
- The `MavenPublishJavaIntegTest` has been turned into a test that runs in multiple configurations, also extending the test coverage for feature variants in general.
- The publishing chapter has been re-organised (again) to clarify that users should add additional variants, or artifacts to existing variants, instead of adding artifacts directly to publications. It also highlights more how this can be done by extending/modifying the existing _java_ component.
- The init task that creates a Gradle build based on a pom has also been updated to use the new feature.
- Basic test coverage for resolving the new variants has been added.

ljacomet commented on subprojects/core/src/main/java/org/gradle/api/internal/artifacts/JavaEcosystemSupport.java:
```diff
@@ -254,6 +261,25 @@ public void execute(CompatibilityCheckDetails<Bundling> details) {
+            String consumerValueName = consumerValue.getName();
+            if (Category.DOCUMENTATION.equals(consumerValueName)) {
+                // In the absence of documentation, everything else can act as 'documentation'
+                details.compatible();
```
If I understand correctly, a consumer that says `org.gradle.category=documentation` will be able to consume all variants, `documentation` or not. But if the producer had two variants, one `documentation`, one `library`, now both will be considered. So you may need a disambiguation rule to make sure we prefer `documentation` when documentation is requested.

ljacomet commented on subprojects/diagnostics/src/integTest/groovy/org/gradle/api/tasks/diagnostics/OutgoingVariantsReportTaskIntegrationTest.groovy:
```diff
@@ -108,6 +108,112 @@ Secondary variants (*)
+          - org.gradle.usage               = java-api
+       - Artifacts
+          - build${File.separator}classes${File.separator}java${File.separator}main (artifactType = java-classes-directory)
+    - Variant : javadoc
```
I wonder about this choice. The `java-runtime` vs `java-api` aspect of the Javadoc kind of depends on the way the `javadoc` task is configured.
Would it not be better to simply have the default `source` and `javadoc` on the `java-runtime` bits? Knowing that users can be more specific if needed.

ljacomet commented on subprojects/docs/src/samples/ivy-publish/java-multi-project/output-ivy.xml:
```diff
@@ -10,16 +10,21 @@
   <dependencies>
     <dependency org="junit" name="junit" rev="4.12" conf="runtime-&gt;default"/>
     <dependency org="org.gradle.sample" name="project2" rev="1.0" conf="runtime-&gt;default"/>
+    <dependency org="junit" name="junit" rev="4.12" conf="runtimeElementsSources-&gt;default"/>
```
I wonder about the choice of doing secondary variants given this. The reason we have dependencies here is because of that relationship between the `sources` and `runtimeElements`.

ljacomet commented on subprojects/plugins/src/integTest/groovy/org/gradle/java/JavaLibraryDocumentationIntegrationTest.groovy:
```diff
@@ -0,0 +1,201 @@
+        }
+    }
+
+    def "javadoJar task is only created when publishing is requested"() {
```
Typo: missing 'c'

jjohannes commented:
@ljacomet I changed the implementation to _primary_ variants without dependencies as we discussed offline. It also feels like the better solution now.

You can see what this changes by only looking at the last commit I added.

Can you have another look at this?
